src/activity_item_handling.cpp

```diff
@@ -7,266 +9,383 @@
-
-    bool ignoring_interruptions = g->u.activity.ignore_trivial;
-    // get whether `drop_target` is a vehicle cargo, then erase the first element
-    bool to_vehicle = g->u.activity.values[0];
-    g->u.activity.values.erase( g->u.activity.values.begin() );
-    tripoint drop_target = get_item_pointers_from_activity( selected_items, item_quantities,
-                           selected_worn_items, worn_item_quantities );
-
-    // Consume the list as long as we don't run out of moves.
-    while( g->u.moves >= 0 && !selected_items.empty() ) {
-        place_item_activity( selected_items, item_quantities,
-                             selected_worn_items, worn_item_quantities,
-                             ( act == ACT_DROP ) ? DROP_NOT_WORN : STASH_NOT_WORN, drop_target, to_vehicle );
-    }
-    while( g->u.moves >= 0 && !selected_worn_items.empty() ) {
-        place_item_activity( selected_items, item_quantities,
-                             selected_worn_items, worn_item_quantities,
-                             ( act == ACT_DROP ) ? DROP_WORN : STASH_WORN, drop_target, to_vehicle );
-    }
-    if( selected_items.empty() && selected_worn_items.empty() ) {
-        // Yay we're done, just exit.
+    if( !debug_mode ) {
         return;
     }
-    // If we make it here load anything left into a new activity.
-    make_drop_activity( act, drop_target, selected_items, item_quantities, selected_worn_items,
-                        worn_item_quantities, ignoring_interruptions, to_vehicle );
+
+    std::string res( "Items ordered to drop:\n" );
+    for( const auto &ait : list ) {
+        res += string_format( "Drop %d %s for %d moves\n",
+                              ait.count, ait.it->display_name( ait.count ).c_str(), ait.consumed_moves );
+    }
+    popup( res, PF_GET_KEY );
+}
+
+std::vector<item> obtain_activity_items( player_activity &act, player &p )
+{
+    std::vector<item> res;
+
+    auto items = reorder_for_dropping( p, convert_to_indexes( act ) );
+
+    debug_drop_list( items );
+
+    while( !items.empty() && ( p.is_npc() || p.moves > 0 || items.front().consumed_moves == 0 ) ) {
+        const auto &ait = items.front();
+
+        p.mod_moves( -ait.consumed_moves );
+
+        if( p.is_worn( *ait.it ) ) {
+            p.takeoff( *ait.it, [ &res ]( const item & tit ) {
+                res.push_back( tit );
+                return true;
+            } );
+        } else if( ait.it->count_by_charges() ) {
+            res.push_back( p.reduce_charges( const_cast<item *>( ait.it ), ait.count ) );
+        } else {
+            res.push_back( p.i_rem( ait.it ) );
+        }
+
+        items.pop_front();
+    }
+    // Avoid tumbling to the ground. Unload cleanly.
+    const int excessive_volume = p.volume_carried() - p.volume_capacity();
+    if( excessive_volume > 0 ) {
+        const auto excess = p.inv.remove_randomly_by_volume( excessive_volume );
+
+        res.reserve( res.size() + excess.size() );
```

Comment:
> Why not just res.insert( res.begin(), excess.begin(), excess.end() ); instead of the five lines?

Good point.


> so I wonder what can actually be moved here.

Probably none, but I want to avoid copying in either case (keeping in mind possible item uids).